keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,372 @@
+
+Basic Azure Load Balancers are free.
+
+Pricing of the Standard Azure Load Balancer is based on load-balancing rules and outbound rules.  There is a flat price up to 5 rules, on top of which every new forwarding rule has additional cost. 
+https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-overview#pricing
+
+A user can ask for an internal Azure Load Balancer via a K8s Service definition that also has the annotation `service.beta.kubernetes.io/azure-load-balancer-internal: true`. There is not any limitation on the usage of mixed service protocols with internal LBs in the Azure CPI.
+
+Summary: Azure already has mixed protocol support for TCP and UDP, and it already affects the bills of the users. The implementation of this feature may require some work in Azure CPI.
+
+#### GCE
+
+The GCE CPI supports both TCP and UDP protocols in Services. Other protocols are not supported.
+
+GCE/GKE creates Network Load Balancers based on the K8s Services with type=LoadBalancer. In GCE there are "forwarding rules" that define how the incoming traffic shall be forwarded to the compute instances. A single forwarding rule can support either TCP or UDP but not both. In order to have both TCP and UDP forwarding rules we have to create separate forwarding rule instances for those. Two or more forwarding rules can share the same external IP if
+- the network load balancer type is External
+- the external IP address is not ephemeral but static
+
+HTTP protocol support: there is a different LB type in GCE for HTTP traffic: HTTP(S) LB. I.e. just like in the case of AWS it is not possible to have e.g. TCP+HTTP or UDP+HTTP behind the same LB/IP address.
+
+Forwarding rule pricing is per rule: there is a flat price up to 5 forwarding rule instances, on top of which every new forwarding rule has additional cost. 
+
+https://cloud.google.com/compute/network-pricing#forwarding_rules_charges
+
+As we can see two different protocols in the same K8s Service definition would result in the creation of 2 forwarding rules in GCP, which has the same fixed price only up to 5 forwarding rule instances, but on top of that every new rule means extra cost.
+
+A user can ask for an internal TCP/UDP Load Balancer via a K8s Service definition that also has the annotation `cloud.google.com/load-balancer-type: "Internal"`. Forwarding rules are also part of the GCE Internal TCP/UDP Load Balancer architecture, but in case of Internal TCP/UDP Load Balancer it is not supported to define different forwarding rules with different protocols for the same IP address. That is, for Services with type=LoadBalancer and with annotation `cloud.google.com/load-balancer-type: "Internal"` this feature would not be supported.
+
+Summary: the implementation of this feature can affect the bills of the users. On the other hand, we can see it from these perspectives also:
+- currently the users cannot have mixed protocols behind the same NLB - it blocks some use cases on GCE
```

Comment:
Possibly link to the GCE users in the issue/PR that are affected by this

Reply:
It is hard to find exact GCE users. I can tell that the one who invented the workaround on GCE is on GCE, but he already has the workaround :) So, it may be not convincing to link that comment here.